attivita/migrations/0018_negazione_automatica_vecchie.py

```diff
@@ -4,38 +4,20 @@
 from datetime import datetime, date, timedelta
 
 from django.db import migrations, models
+from anagrafica.permessi.incarichi import INCARICO_GESTIONE_ATTIVITA_PARTECIPANTI
 
 
 def nega_partecipazioni_vecchie(apps, schema_editor):
 
-    Partecipazione = apps.get_model("attivita", "Partecipazione")
     Autorizzazione = apps.get_model("base", "Autorizzazione")
 
-    partecipzioni_pendenti = Partecipazione.objects.filter(confermata=False, ritirata=False, automatica=False)
-    totale = partecipzioni_pendenti.count()
-    print("  => 0018 trovate %d partecipazioni di attivita pendenti" % (totale))
-
-    autorizzazioni_con_scadenza = 0
-    autorizzazioni_senza_scadenza = 0
-    for partecipazione in partecipzioni_pendenti:
-        # Autorizzazioni senza esito della partecipazione pendente
-        autorizzazioni_partecipazione = Autorizzazione.objects.filter(oggetto_id=partecipazione.pk, concessa=None)
-        for autorizzazione in autorizzazioni_partecipazione:
-            # Se l'autorizzazione ha una scadenza
-            if autorizzazione.scadenza:
-                # Imposta negazione automatica per autorizzazione con scadenza
-                if(autorizzazione.tipo_gestione != 'N'):
-                    autorizzazione.tipo_gestione = 'N'
-                    autorizzazione.save()
-                    autorizzazioni_con_scadenza += 1
-            else:
-                # Imposta negazione automatica e scadenza 30gg per autorizzazione senza scadenza
-                autorizzazione.tipo_gestione = 'N'
-                autorizzazione.scadenza = autorizzazione.creazione + timedelta(days=30)
-                autorizzazione.save()
-                autorizzazioni_senza_scadenza += 1
-
-    print("  ==> 0018 modificate %d autorizzazioni con scadenza e %d autorizzazioni senza scadenza" % (autorizzazioni_con_scadenza, autorizzazioni_senza_scadenza))
+    # Autorizzazioni attività di partecipazione, senza esito ed escludendo quelle già negate.
+    autorizzazioni = Autorizzazione.objects.filter(concessa__isnull=True,
+                                                   destinatario_ruolo=INCARICO_GESTIONE_ATTIVITA_PARTECIPANTI
```

Comment:
Nelle migrazioni (e solo nelle migrazioni) cerca di evitare l'uso delle costanti. Questo e' anche parte del motivo per cui Django ti impedisce di ottenere costanti dai modelli e ti obbliga di passare da `apps.get_model` (se cerchi di ottenere una costante su di un modello ritornato da quella funzione, come ad esempio `Autorizzazione.AUTOMATICA` vedrai che questo fallira'). E' una feature, non un bug!

In Django, le migrazioni devono essere indipendenti dalla versione dei tuoi modelli e, piu' in generale, di qualunque codice oltre che le migrazioni stesse.

Se un giorno il valore di quella costante dovesse cambiare, le migrazioni che fanno riferimento ad una vecchia versione del database *non* dovrebbero cambiare di conseguenza. `apps.get_model` ti ritorna una versione del modello che e' storicamente uguale a quella che era presente quando la migrazione e' stata creata.

Spesso, si fa riferimento a questo concetto definendolo il "congelamento" dell'ORM. Puoi leggerne una spiegazione migliore della mia alla pagina [ORM Freezing](https://south.readthedocs.io/en/latest/) della documentazione di Django South, il progetto che poi e' stato integrato in Django stesso come gestore delle migrazioni.